godot-core/src/builtin/script.rs

```diff
@@ -0,0 +1,638 @@
+
+        let script = borrow_instance(instance).get_script();
+
+        transfer_script_to_godot(script)
+    }
+
+    pub(super) unsafe extern "C" fn is_placeholder_func<T: ScriptInstance>(
+        p_instance: GDExtensionScriptInstanceDataPtr,
+    ) -> u8 {
+        let instance = instance_data_as_script_instance::<T>(p_instance);
+
+        let is_placeholder = borrow_instance(instance).is_placeholder();
+
+        transfer_bool_to_godot(is_placeholder)
+    }
+
+    pub(super) unsafe extern "C" fn has_method_func<T: ScriptInstance>(
+        p_instance: GDExtensionScriptInstanceDataPtr,
+        p_method: GDExtensionConstStringNamePtr,
+    ) -> u8 {
+        let instance = instance_data_as_script_instance::<T>(p_instance);
+        let method = transfer_string_name_from_godot(p_method);
+
+        let has_method = borrow_instance(instance).has_method(method.clone());
+
+        transfer_bool_to_godot(has_method)
+    }
+
+    pub(super) unsafe extern "C" fn free_method_list_func<T: ScriptInstance>(
+        p_instance: GDExtensionScriptInstanceDataPtr,
+        p_method_info: *const GDExtensionMethodInfo,
+    ) {
+        let instance = instance_data_as_script_instance::<T>(p_instance);
+        let length = borrow_instance(instance).method_list().len();
+
+        let vec = transfer_ptr_list_from_godot(p_method_info, length);
+
+        vec.into_iter().for_each(|method_info| {
+            transfer_ptr_list_from_godot(
+                method_info.arguments,
+                method_info.argument_count as usize,
+            );
+            transfer_ptr_list_from_godot(
+                method_info.default_arguments,
+                method_info.default_argument_count as usize,
+            );
+        })
+    }
+
+    pub(super) unsafe extern "C" fn get_property_type_func<T: ScriptInstance>(
+        p_instance: GDExtensionScriptInstanceDataPtr,
+        p_name: GDExtensionConstStringNamePtr,
+        r_is_valid: *mut GDExtensionBool,
+    ) -> GDExtensionVariantType {
+        let instance = instance_data_as_script_instance::<T>(p_instance);
+        let name = transfer_string_name_from_godot(p_name);
+
+        let result = borrow_instance(instance).get_property_type(name.clone());
+
+        *r_is_valid = transfer_bool_to_godot(true);
+        result.sys()
+    }
+
+    pub(super) unsafe extern "C" fn to_string_func<T: ScriptInstance>(
+        p_instance: GDExtensionScriptInstanceDataPtr,
+        r_is_valid: *mut GDExtensionBool,
+        r_str: GDExtensionStringPtr,
+    ) {
+        let instance = instance_data_as_script_instance::<T>(p_instance);
+
+        let Ok(inner) = try_borrow_instance(instance) else {
+            // to_string of a script instance can be called when calling to_string
+            // on the owning base object. In this case we pretend like we can't
+            // handle the call and leave r_is_valid at it's default value of false.
+
+            return;
+        };
```

Comment:
Why did you choose this behavior?
What are the implications if someone wants to get the string representation?

Reply:
Thanks a lot for the elaborate description. Could you maybe add this to the code itself?
Also the point about potential future changes can be quite helpful 👍

Reply:
Sorry i didn't state a concrete question before. Do you agree with my take of tasking the implementor with using interior mutability them self and changing the trait to only take immutable references? 

If this approach is liked by everyone, I will make the change. 

(I've now also started working on tests btw)